File: crates/compiler/frontend/typescript/src/markers.ts
// Public authoring markers for typed Agent source.
//
// These are compile-time declarations resolved by imported symbol identity and
// type. They declare schemas, bindings, and handlers; they never execute the
// authored body to discover the graph, and they carry no credential, grant,
// endpoint, or runtime object.

import { recordDeclaration } from "./declared.js";
import {
  CAPABILITY_DISPLAY_NAME_REJECTED,
  CAPABILITY_REF_NOT_EXACT,
  type DiagnosticCode,
  EVENT_NOT_TYPED,
  EVENT_WAIT_OUTSIDE_BODY,
  MODEL_DISPLAY_NAME_REJECTED,
  SKILL_ENTRY_PATH_NOT_CANONICAL,
  SKILL_ID_NOT_EXACT,
  SKILL_INSTRUCTIONS_OVERLONG,
  SKILL_LOAD_OUTSIDE_BODY,
  SKILL_SOURCE_AMBIGUOUS,
  SKILL_SOURCE_MISSING,
  TOOL_DISPLAY_NAME_REJECTED,
  TOOL_REF_NOT_CAPABILITY,
} from "./generated/diagnostics.js";
import {
  SKILL_INSTRUCTION_KIND_ENTRY,
  SKILL_INSTRUCTION_KIND_INLINE,
} from "./generated/frontend-graph.js";
import type { SkillInstructionSource } from "./generated/frontend-records.js";
import {
  PERMISSION_DECISIONS,
  type Permission,
} from "./generated/permissions.js";
// The generated catalogue's `CapabilityId` is the closed set of builtin ids; the
// packaging surface's `CapabilityId` is the object a handler declaration hands
// back. Two different concepts wearing one name, so each is renamed to the thing
// it actually is at the point where both meet.
import type { CapabilityId as BuiltinCapabilityId } from "./generated/capabilities.js";
import { BUILTIN_CAPABILITIES } from "./generated/capabilities.js";

// Capture the intrinsic before authored code can replace the global method.
// Marker records are inputs to static capture, so their fields must remain
// stable after the declaration factory returns.
const objectFreeze = Object.freeze.bind(Object);

/**
 * The ceiling the skill-reading capability enforces on a body it loads. An
 * inline skill is the same trusted context landing in the same model window, so
 * it is refused here rather than at execution.
 */
const MAX_INSTRUCTION_BYTES = 128 * 1024;

const FORBIDDEN_DISPLAY_NAMES = new Set(["default", "model.default", "support", "search-web", ""]);

/**
 * The reference a shipped handler declaration hands back. The id and the
 * implementation arrive as one object, so there is no second place to spell the
 * id and nothing for the two spellings to disagree about.
 */
export type ShippedCapabilityReference = { readonly capabilityId: string };

/**
 * A Capability reference is a builtin id from the generated catalogue or the
 * handler that implements one. Nothing else: an invented bare string is not a
 * member of either arm, so referencing a Capability nobody implements is a type
 * error rather than a check some later pass has to remember to run.
 */
export type CapabilityReference = BuiltinCapabilityId | ShippedCapabilityReference;

function capabilityIdOf(reference: CapabilityReference): string {
  return typeof reference === "string" ? reference : reference?.capabilityId;
}

function rejectDisplayName(value: unknown, marker: string, code: DiagnosticCode): void {
  if (typeof value !== "string" || FORBIDDEN_DISPLAY_NAMES.has(value)) {
    throw new Error(
      `${code}: ${marker} accepts an exact typed reference, not a display name '${value}'`,
    );
  }
}

/**
 * Refuse a Capability reference that neither catalogue nor package mints.
 *
 * The type union already settles this for TypeScript source. JavaScript source
 * is authored against the same surface with no type-checker between it and the
 * frontend, so the closed set is settled here too: a bare string is admitted
 * only when the generated catalogue mints it, and any other reference has to be
 * the object a handler declaration handed back.
 */
function rejectUnmintedCapability(
  reference: CapabilityReference,
  marker: string,
  code: DiagnosticCode,
): void {
  if (typeof reference !== "string") {
    return;
  }
  if (!(BUILTIN_CAPABILITIES as readonly string[]).includes(reference)) {
    throw new Error(
      `${code}: ${marker} accepts a builtin catalogue id or the handler ` +
        `declaration that implements one, not '${reference}'`,
    );
  }
}

// A binding is authored as a call inside the Agent body; at the type level it is
// a callable so ordinary source type-checks, while the frontend resolves the call
// statically rather than executing it.
export type ModelBinding<I = unknown, O = unknown> = {
  readonly kind: "model_binding";
  readonly targetRef: string;
  readonly inputTypeRef: string;
  readonly outputTypeRef: string;
  (request: I): Promise<O>;
};

export type ToolBinding<I = unknown, O = unknown> = {
  readonly kind: "tool_binding";
  readonly targetRef: string;
  readonly inputTypeRef: string;
  readonly outputTypeRef: string;
  readonly permission?: Permission;
  (args: I): Promise<O>;
};

export type CapabilityBinding<I = unknown, O = unknown> = {
  readonly kind: "capability_binding";
  readonly targetRef: string;
  readonly inputTypeRef: string;
  readonly outputTypeRef: string;
  readonly permission?: Permission;
  (args: I): Promise<O>;
};

export type EventTypeBinding<T = unknown> = {
  readonly kind: "event_type";
  readonly typeRef: string;
  readonly targetRef: string;
  wait(): Promise<T>;
};

export type ContextSchema = {
  readonly kind: "context";
  readonly typeRef: string;
  readonly defaultPresent: boolean;
};

/**
 * Where a declared skill's instructions live. Exactly one field is stated: the
 * two are the two routes an edit takes to the artifact digest, and stating both
 * is a contradiction about where the instructions are.
 */
export type SkillSource = {
  readonly entry?: string;
  readonly text?: string;
};

/**
 * A declared Agent Skill. Loading one is an ordinary Capability invocation, not
 * a construct of its own: `await skill.load()` records a `capability.invoke` on
 * the skill-reading capability, so the authority to read the instructions is
 * declared in the artifact like any other.
 */
export type SkillBinding = {
  readonly kind: "skill";
  readonly skillId: string;
  readonly instructionSource: SkillInstructionSource;
  load(): Promise<string>;
};

/** The one package path the folder contract recognizes for this skill. */
function skillEntryPath(skillId: string): string {
  return `skills/${skillId}/SKILL.md`;
}

/** Derive a deterministic source-identity digest for static declarations. */
export function stableDigest(value: string): string {
  const bytes = new TextEncoder().encode(value);
  const paddedLength = ((bytes.length + 9 + 63) >> 6) << 6;
  const padded = new Uint8Array(paddedLength);
  padded.set(bytes);
  padded[bytes.length] = 0x80;
  const bitLength = BigInt(bytes.length) * 8n;
  for (let index = 0; index < 8; index += 1) {
    padded[paddedLength - 1 - index] = Number((bitLength >> BigInt(index * 8)) & 0xffn);
  }

  const state = [
    0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a,
    0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19,
  ];
  const words = new Uint32Array(64);
  for (let offset = 0; offset < padded.length; offset += 64) {
    for (let index = 0; index < 16; index += 1) {
      const position = offset + index * 4;
      words[index] = (
        (padded[position] << 24) |
        (padded[position + 1] << 16) |
        (padded[position + 2] << 8) |
        padded[position + 3]
      ) >>> 0;
    }
    for (let index = 16; index < 64; index += 1) {
      const first = words[index - 15];
      const second = words[index - 2];
      const smallSigma0 = rotateRight(first, 7) ^ rotateRight(first, 18) ^ (first >>> 3);
      const smallSigma1 = rotateRight(second, 17) ^ rotateRight(second, 19) ^ (second >>> 10);
      words[index] = (words[index - 16] + smallSigma0 + words[index - 7] + smallSigma1) >>> 0;
    }

    let [a, b, c, d, e, f, g, h] = state;
    for (let index = 0; index < 64; index += 1) {
      const bigSigma1 = rotateRight(e, 6) ^ rotateRight(e, 11) ^ rotateRight(e, 25);
      const choice = (e & f) ^ (~e & g);
      const temp1 = (h + bigSigma1 + choice + SHA256_ROUND_CONSTANTS[index] + words[index]) >>> 0;
      const bigSigma0 = rotateRight(a, 2) ^ rotateRight(a, 13) ^ rotateRight(a, 22);
      const majority = (a & b) ^ (a & c) ^ (b & c);
      const temp2 = (bigSigma0 + majority) >>> 0;
      h = g;
      g = f;
      f = e;
      e = (d + temp1) >>> 0;
      d = c;
      c = b;
      b = a;
      a = (temp1 + temp2) >>> 0;
    }
    state[0] = (state[0] + a) >>> 0;
    state[1] = (state[1] + b) >>> 0;
    state[2] = (state[2] + c) >>> 0;
    state[3] = (state[3] + d) >>> 0;
    state[4] = (state[4] + e) >>> 0;
    state[5] = (state[5] + f) >>> 0;
    state[6] = (state[6] + g) >>> 0;
    state[7] = (state[7] + h) >>> 0;
  }
  return `sha256:${state.map((word) => word.toString(16).padStart(8, "0")).join("")}`;
}

function rotateRight(value: number, amount: number): number {
  return (value >>> amount) | (value << (32 - amount));
}

const SHA256_ROUND_CONSTANTS = [
  0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
  0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
  0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
  0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
  0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
  0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
  0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
  0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
];

function uncallable(marker: string): never {
  throw new Error(`a ${marker} is invoked inside a compiled Agent body`);
}

/** Keep one declaration record immutable after it enters the module registry. */
function freezeDeclaration<T extends object>(declaration: T): T {
  return objectFreeze(declaration);
}

/** Copy a permission into an immutable plain record before capture can read it. */
function freezePermission(permission: Permission | undefined): Permission | undefined {
  if (permission === undefined) {
    return undefined;
  }
  const snapshot: Permission = {
    decision: permission.decision,
    ...(permission.reason === undefined ? {} : { reason: permission.reason }),
  };
  return freezeDeclaration(snapshot);
}

export function Model<Input, Output>(ref: string): ModelBinding<Input, Output> {
  rejectDisplayName(ref, "Model", MODEL_DISPLAY_NAME_REJECTED);
  const binding = () => uncallable("Model");
  return recordDeclaration(freezeDeclaration(Object.assign(binding, {
    kind: "model_binding" as const,
    targetRef: ref,
    inputTypeRef: "ModelRequest",
    outputTypeRef: "ModelResponse",
  })));
}

export function Tool<Input, Output>(
  capabilityRef: CapabilityReference,
  options: { readonly permission?: Permission } = {},
): ToolBinding<Input, Output> {
  const binding = () => uncallable("Tool");
  const targetRef = capabilityIdOf(capabilityRef);
  rejectDisplayName(targetRef, "Tool", TOOL_DISPLAY_NAME_REJECTED);
  rejectUnmintedCapability(capabilityRef, "Tool", TOOL_REF_NOT_CAPABILITY);
  rejectInvalidPermission(options.permission, "Tool");
  return recordDeclaration(freezeDeclaration(Object.assign(binding, {
    kind: "tool_binding" as const,
    targetRef,
    inputTypeRef: "ToolInput",
    outputTypeRef: "ToolOutput",
    permission: freezePermission(options.permission),
  })));
}

export function Capability<Input, Output>(
  ref: CapabilityReference,
  options: { readonly permission?: Permission } = {},
): CapabilityBinding<Input, Output> {
  const binding = () => uncallable("Capability");
  const targetRef = capabilityIdOf(ref);
  rejectDisplayName(targetRef, "Capability", CAPABILITY_DISPLAY_NAME_REJECTED);
  rejectUnmintedCapability(ref, "Capability", CAPABILITY_REF_NOT_EXACT);
  rejectInvalidPermission(options.permission, "Capability");
  return recordDeclaration(freezeDeclaration(Object.assign(binding, {
    kind: "capability_binding" as const,
    targetRef,
    inputTypeRef: "CapabilityInput",
    outputTypeRef: "CapabilityOutput",
    permission: freezePermission(options.permission),
  })));
}

/** Keep JavaScript callers on the generated closed permission vocabulary. */
function rejectInvalidPermission(
  permission: Permission | undefined,
  marker: string,
): void {
  if (permission === undefined) {
    return;
  }
  const candidate = permission as unknown as {
    decision?: unknown;
    reason?: unknown;
  };
  const decision = candidate?.decision;
  if (
    (typeof permission !== "object" && typeof permission !== "function") ||
    permission === null ||
    typeof decision !== "string" ||
    !(PERMISSION_DECISIONS as readonly string[]).includes(decision)
  ) {
    throw new TypeError(
      `${marker} permission is one of the generated Allow, Ask, or Deny markers`,
    );
  }
  if (
    candidate.reason !== undefined &&
    (typeof candidate.reason !== "string" || candidate.reason.length === 0)
  ) {
    throw new TypeError(`${marker} permission reason must be a non-empty string`);
  }
}

export function Event<Payload>(ref: string): EventTypeBinding<Payload> {
  rejectDisplayName(ref, "Event", EVENT_NOT_TYPED);
  return recordDeclaration(freezeDeclaration({
    kind: "event_type" as const,
    typeRef: "Event",
    targetRef: ref,
    wait: (): never => {
      throw new Error(
        `${EVENT_WAIT_OUTSIDE_BODY}: an Event is awaited inside a compiled Agent body`,
      );
    },
  }));
}

/**
 * Declare one Agent Skill, its instructions carried one of two ways.
 *
 * `entry` names the package file holding them, which the carrying package's
 * integrity chain hashes. `text` writes them here, inside the source bundle the
 * artifact digest already covers. Exactly one is stated.
 */
export function Skill(skillId: string, source: SkillSource): SkillBinding {
  if (typeof skillId !== "string" || FORBIDDEN_DISPLAY_NAMES.has(skillId)) {
    throw new Error(
      `${SKILL_ID_NOT_EXACT}: Skill accepts an exact typed reference, not a display name '${skillId}'`,
    );
  }
  const entry = source?.entry;
  const text = source?.text;
  if (entry !== undefined && text !== undefined) {
    throw new Error(
      `${SKILL_SOURCE_AMBIGUOUS}: Skill '${skillId}' states both a package entry and inline text; instructions live in one place`,
    );
  }
  if (entry === undefined && text === undefined) {
    throw new Error(
      `${SKILL_SOURCE_MISSING}: Skill '${skillId}' states neither a package entry nor inline text`,
    );
  }
  let instructionSource: SkillInstructionSource;
  if (text !== undefined) {
    if (text === "" || new TextEncoder().encode(text).length > MAX_INSTRUCTION_BYTES) {
      throw new Error(
        `${SKILL_INSTRUCTIONS_OVERLONG}: Skill '${skillId}' states an empty or oversized body; a loaded skill is at most ${MAX_INSTRUCTION_BYTES} bytes`,
      );
    }
    instructionSource = freezeDeclaration({ kind: SKILL_INSTRUCTION_KIND_INLINE, text });
  } else {
    const expected = skillEntryPath(skillId);
    if (entry !== expected) {
      throw new Error(
        `${SKILL_ENTRY_PATH_NOT_CANONICAL}: Skill '${skillId}' carries its instructions at '${expected}', not '${entry}'`,
      );
    }
    instructionSource = freezeDeclaration({ kind: SKILL_INSTRUCTION_KIND_ENTRY, path: expected });
  }
  return recordDeclaration(freezeDeclaration({
    kind: "skill" as const,
    skillId,
    instructionSource,
    load: (): never => {
      throw new Error(
        `${SKILL_LOAD_OUTSIDE_BODY}: a Skill is loaded inside a compiled Agent body`,
      );
    },
  }));
}

export function Context<Schema>(initial?: Schema): ContextSchema {
  return recordDeclaration(freezeDeclaration({
    kind: "context" as const,
    // The Context type identity is read off `Context<Schema>` in the authored
    // source, exactly as Python reads it off the decorated class.
    typeRef: "Context",
    defaultPresent: initial !== undefined,
  }));
}

File: crates/compiler/frontend/python/apxm_program/_markers.py
"""Public authoring markers for typed Agent source.

These are compile-time declarations resolved by imported symbol identity and
type. They declare schemas, bindings, and handlers; they never execute the
authored body to discover the graph, and they carry no credential, grant,
endpoint, or runtime object.
"""

from __future__ import annotations

from typing import Any, Awaitable, Generic, NamedTuple, Optional, TypeVar

from ._generated.capabilities import BUILTIN_CAPABILITIES
from ._host_capabilities import (
    HOST_CAPABILITY_REF_PREFIX,
    declared_host_capabilities,
    is_declared_host_capability,
)
from ._generated.diagnostics import (
    CAPABILITY_DISPLAY_NAME_REJECTED,
    CAPABILITY_REF_NOT_EXACT,
    CONTEXT_NOT_TYPED,
    DiagnosticCode,
    EVENT_NOT_TYPED,
    EVENT_WAIT_OUTSIDE_BODY,
    MODEL_DISPLAY_NAME_REJECTED,
    MODEL_UNTYPED_SCHEMA,
    SKILL_ENTRY_PATH_NOT_CANONICAL,
    SKILL_ID_NOT_EXACT,
    SKILL_INSTRUCTIONS_OVERLONG,
    SKILL_LOAD_OUTSIDE_BODY,
    SKILL_SOURCE_AMBIGUOUS,
    SKILL_SOURCE_MISSING,
    TOOL_DISPLAY_NAME_REJECTED,
    TOOL_REF_NOT_CAPABILITY,
)
from ._generated.frontend_graph import (
    SKILL_INSTRUCTION_KIND_ENTRY,
    SKILL_INSTRUCTION_KIND_INLINE,
)
from ._generated.frontend_records import (
    SkillInstructionSource,
)
from ._generated.permissions import PERMISSION_DECISIONS, Permission
from .handlers import CapabilityId

#: The ceiling the skill-reading capability enforces on a body it loads. An
#: inline skill is the same trusted context landing in the same model window,
#: so it is refused here rather than at execution.
MAX_INSTRUCTION_BYTES = 128 * 1024

I = TypeVar("I")
O = TypeVar("O")
T = TypeVar("T")


def _type_name(annotation: Any, fallback: str) -> str:
    if annotation is None:
        return fallback
    if isinstance(annotation, str):
        return annotation
    return getattr(annotation, "__name__", str(annotation))


class ModelBinding(NamedTuple):
    """An exact typed model target. Calling it records a model invocation."""

    target_ref: str
    input_type_ref: str = "ModelRequest"
    output_type_ref: str = "ModelResponse"

    def __call__(self, *args: Any, **kwargs: Any) -> Any:  # pragma: no cover
        raise RuntimeError("a Model is invoked inside a compiled Agent body")


class ToolBinding(NamedTuple):
    """A static model-callable capability reference."""

    target_ref: str
    input_type_ref: str = "ToolInput"
    output_type_ref: str = "ToolOutput"
    permission: Optional[Permission] = None

    def __call__(self, *args: Any, **kwargs: Any) -> Any:  # pragma: no cover
        raise RuntimeError("a Tool is invoked inside a compiled Agent body")


class CapabilityBinding(NamedTuple):
    """A static executable capability reference that is not a Tool."""

    target_ref: str
    input_type_ref: str = "CapabilityInput"
    output_type_ref: str = "CapabilityOutput"
    permission: Optional[Permission] = None
    output_annotation: Any = None

    def __call__(self, *args: Any, **kwargs: Any) -> Any:  # pragma: no cover
        raise RuntimeError("a Capability is invoked inside a compiled Agent body")


class EventRef(Generic[T]):
    """An opaque admitted reservation, never constructed by program source."""

    __slots__ = ()
    event_id: str
    generation: int

    def __new__(cls) -> EventRef[T]:
        raise TypeError("EventRef values enter through admitted input or Capability results")


class EventType(NamedTuple, Generic[T]):
    """A typed durable event reference whose wait records an event wait."""

    type_ref: str
    target_ref: str
    payload_type: type[T]

    def wait(self, reference: EventRef[T]) -> Awaitable[T]:  # pragma: no cover
        raise RuntimeError(
            f"{EVENT_WAIT_OUTSIDE_BODY}: an Event is awaited inside a compiled Agent body"
        )


class _SkillEntrySource(NamedTuple):
    """Immutable internal representation of a package-carried skill source."""

    kind: str
    path: str


class _SkillInlineSource(NamedTuple):
    """Immutable internal representation of an inline skill source."""

    kind: str
    text: str


class SkillDecl(NamedTuple):
    """A declared Agent Skill: instructions plus where they live.

    Loading one is an ordinary Capability invocation, not a construct of its
    own: ``await skill.load()`` records a ``capability.invoke`` on the
    skill-reading capability, so the authority to read the instructions is
    declared in the artifact like any other.
    """

    skill_id: str
    instruction_source: SkillInstructionSource

    def load(self, *args: Any, **kwargs: Any) -> Any:  # pragma: no cover
        raise RuntimeError(
            f"{SKILL_LOAD_OUTSIDE_BODY}: a Skill is loaded inside a compiled Agent body"
        )


class ContextSchema(tuple, Generic[T]):
    """The typed initial and persistent Program Context schema."""

    __slots__ = ()

    def __new__(cls, type_ref: str, default_present: bool, schema_type: type[T]) -> ContextSchema[T]:
        return tuple.__new__(cls, (type_ref, default_present, schema_type))

    @property
    def type_ref(self) -> str:
        return self[0]

    @property
    def default_present(self) -> bool:
        return self[1]

    @property
    def schema_type(self) -> type[T]:
        return self[2]


class _FrozenFactoryType(type):
    """Keep marker factory methods immutable after the frontend loads."""

    def __setattr__(cls, name: str, value: Any) -> None:
        raise AttributeError("authoring marker factories are immutable after import")


class _TypedFactory(metaclass=_FrozenFactoryType):
    """A marker whose type parameters are supplied by subscription.

    ``_type_parameters`` names them in order. It is the Python spelling of the
    same arity a TypeScript marker states as ``<Input, Output>``, which is what
    lets the surface conformance gate compare the two by argument shape rather
    than by trusting that a subscript exists.

    Subscription returns a *new* factory rather than the module singleton, so a
    marker can tell a declaration that stated its types from one that did not.
    """

    _type_parameters: tuple[str, ...] = ()

    def __init__(self, typed: bool = False) -> None:
        self._typed = typed
        self._type_arguments: tuple[Any, ...] = ()

    def __getitem__(self, types: Any) -> "_TypedFactory":
        supplied = types if isinstance(types, tuple) else (types,)
        if len(supplied) != len(self._type_parameters):
            expected = ", ".join(self._type_parameters)
            raise TypeError(f"{type(self).__name__} takes [{expected}]")
        factory = type(self)(typed=True)
        factory._type_arguments = supplied
        return factory


class _ModelFactory(_TypedFactory):
    _type_parameters = ("input", "output")

    def __call__(self, ref: str) -> ModelBinding:
        _require_exact_reference(ref, "Model", MODEL_DISPLAY_NAME_REJECTED)
        if not self._typed:
            raise ValueError(
                f"{MODEL_UNTYPED_SCHEMA}: Model states the request and response it "
                "carries as Model[Input, Output]"
            )
        return ModelBinding(target_ref=ref)


class _ToolFactory(_TypedFactory):
    _type_parameters = ("input", "output")

    def __call__(
        self, capability_ref: str, *, permission: Optional[Permission] = None
    ) -> ToolBinding:
        _require_exact_reference(capability_ref, "Tool", TOOL_DISPLAY_NAME_REJECTED)
        _require_capability_reference(capability_ref, "Tool", TOOL_REF_NOT_CAPABILITY)
        _require_permission(permission, "Tool")
        return ToolBinding(target_ref=capability_ref, permission=permission)


class _CapabilityFactory(_TypedFactory):
    _type_parameters = ("input", "output")

    def __call__(
        self, ref: str, *, permission: Optional[Permission] = None
    ) -> CapabilityBinding:
        _require_exact_reference(ref, "Capability", CAPABILITY_DISPLAY_NAME_REJECTED)
        _require_capability_reference(ref, "Capability", CAPABILITY_REF_NOT_EXACT)
        _require_permission(permission, "Capability")
        return CapabilityBinding(target_ref=ref, permission=permission, output_annotation=self._type_arguments[1] if self._typed else None)


class _TypedEventFactory(Generic[T]):
    """Binds a static Event payload type before its exact event reference."""

    def __init__(self, payload_type: type[T]) -> None:
        self._payload_type = payload_type

    def __call__(self, ref: str) -> EventType[T]:
        _require_exact_reference(ref, "Event", EVENT_NOT_TYPED)
        return EventType(type_ref=_type_name(self._payload_type, "Event"), target_ref=ref, payload_type=self._payload_type)


class _EventFactory(_TypedFactory):
    _type_parameters = ("payload",)

    def __getitem__(self, types: type[T]) -> _TypedEventFactory[T]:
        super().__getitem__(types)
        return _TypedEventFactory(types)

    def __call__(self, ref: str) -> EventType:
        _require_exact_reference(ref, "Event", EVENT_NOT_TYPED)
        raise TypeError(f"{EVENT_NOT_TYPED}: Event requires an explicit finite payload type")


def Context(schema: type[T]) -> ContextSchema[T]:
    """Declare a typed Program Context schema from one typed class."""
    if not isinstance(schema, type):
        raise TypeError(f"{CONTEXT_NOT_TYPED}: Context decorates one typed class")
    fields = [name for name in getattr(schema, "__annotations__", {}) if not name.startswith("__")]
    default_present = all(name in vars(schema) for name in fields)
    return ContextSchema(type_ref=schema.__name__, default_present=default_present, schema_type=schema)


def Skill(
    skill_id: str, *, entry: Optional[str] = None, text: Optional[str] = None
) -> SkillDecl:
    """Declare one Agent Skill, its instructions carried one of two ways.

    ``entry`` names the package file holding them, which the carrying package's
    integrity chain hashes. ``text`` writes them here, inside the source bundle
    the artifact digest already covers. They are the two routes an edit takes to
    the artifact digest, so exactly one is stated: both is a contradiction about
    where the instructions live, and neither leaves the skill with no body.
    """
    if not _is_exact_reference(skill_id):
        raise ValueError(
            f"{SKILL_ID_NOT_EXACT}: Skill accepts an exact typed reference, not a "
            f"display name '{skill_id}'"
        )
    if entry is not None and text is not None:
        raise ValueError(
            f"{SKILL_SOURCE_AMBIGUOUS}: Skill '{skill_id}' states both a package "
            "entry and inline text; instructions live in one place"
        )
    if entry is None and text is None:
        raise ValueError(
            f"{SKILL_SOURCE_MISSING}: Skill '{skill_id}' states neither a package "
            "entry nor inline text"
        )
    source: SkillInstructionSource
    if text is not None:
        if not text or len(text.encode("utf-8")) > MAX_INSTRUCTION_BYTES:
            raise ValueError(
                f"{SKILL_INSTRUCTIONS_OVERLONG}: Skill '{skill_id}' states an empty "
                f"or oversized body; a loaded skill is at most {MAX_INSTRUCTION_BYTES} bytes"
            )
        source = _SkillInlineSource(kind=SKILL_INSTRUCTION_KIND_INLINE, text=text)
    else:
        expected = _skill_entry_path(skill_id)
        if entry != expected:
            raise ValueError(
                f"{SKILL_ENTRY_PATH_NOT_CANONICAL}: Skill '{skill_id}' carries its "
                f"instructions at '{expected}', not '{entry}'"
            )
        source = _SkillEntrySource(kind=SKILL_INSTRUCTION_KIND_ENTRY, path=expected)
    return SkillDecl(skill_id=skill_id, instruction_source=source)


def _skill_entry_path(skill_id: str) -> str:
    """The one package path the folder contract recognizes for this skill."""
    return f"skills/{skill_id}/SKILL.md"


def _is_exact_reference(value: Any) -> bool:
    forbidden = {"default", "model.default", "support", "search-web", ""}
    return isinstance(value, str) and value not in forbidden


def _require_exact_reference(value: Any, marker: str, code: DiagnosticCode) -> None:
    if not _is_exact_reference(value):
        raise ValueError(
            f"{code}: {marker} accepts an exact typed reference, not a "
            f"display name '{value}'"
        )


def _require_capability_reference(
    value: Any, marker: str, code: DiagnosticCode
) -> None:
    """Refuse a Capability reference that neither catalogue nor package mints.

    TypeScript narrows this to a union and the compiler settles it; Python has
    no type-checker in the build, so the same closed set is settled here, where
    the binding is constructed. Two arms are admitted, and they are the two
    places a Capability can come from:

    * a builtin id the generated catalogue mints, which the compiler's builtin
      allowlist admits without any registration;
    * a ``host:<id>`` reference the package's ``agent.toml`` declares as
      ``[[capabilities.host]]``, which the runtime never executes and the
      embedding host answers; and
    * the :class:`CapabilityId` a ``capability(...)`` declaration returns, which
      carries the implementation it names.

    ``CapabilityId`` subclasses ``str`` so an author can still print it and
    compare it to the id it spells, but ``isinstance`` — not ``==`` — is what is
    asked here, so an equal bare string is not mistaken for the declaration that
    would have implemented it.
    """
    if isinstance(value, CapabilityId) or value in BUILTIN_CAPABILITIES:
        return
    if isinstance(value, str) and value.startswith(HOST_CAPABILITY_REF_PREFIX):
        if is_declared_host_capability(value):
            return
        declared = declared_host_capabilities()
        raise ValueError(
            f"{code}: {marker} names the host-fulfilled Capability '{value}', "
            "which the package manifest does not declare as "
            "[[capabilities.host]]. "
            + (
                "This package declares none"
                if not declared
                else "It declares " + ", ".join(declared)
            )
        )
    raise ValueError(
        f"{code}: {marker} accepts a builtin catalogue id, a host-fulfilled "
        f"'host:' reference the manifest declares, or the handler declaration "
        f"that implements one, not '{value}'"
    )


def _require_permission(value: Any, marker: str) -> None:
    """Keep runtime callers on the generated closed permission vocabulary.

    Python type annotations do not run when a package is imported, and a
    ``Permission`` dataclass can therefore be constructed with an invalid
    decision or an empty reason. Refuse those values while the authored marker
    is built so malformed ``requested_permission`` records never reach the
    FrontendGraph bridge.
    """
    if value is None:
        return
    if not isinstance(value, Permission):
        raise TypeError(
            f"{marker} permission is one of the generated Allow, Ask, or Deny "
            f"markers, not {value!r}"
        )
    if value.decision not in PERMISSION_DECISIONS:
        raise ValueError(
            f"{marker} permission decision must be one of "
            f"{', '.join(PERMISSION_DECISIONS)}, not {value.decision!r}"
        )
    if value.reason is not None and (
        not isinstance(value.reason, str) or not value.reason
    ):
        raise ValueError(f"{marker} permission reason must be a non-empty string")


Model = _ModelFactory()
Tool = _ToolFactory()
Capability = _CapabilityFactory()
Event = _EventFactory()
